File: BOJ/python3/19-1 Sogang ACM-ICPC Team Application (Silver)/9012.py
'''
괄호
https://www.acmicpc.net/problem/9012
'''
T = int(input())  # 테스트케이스
string = []
stack = []
for _ in range(T):
    string = list(input())  # 괄호 문자열 입력

    for s in string:
        if s == '(':
            stack.append(s)
        else:
            top = len(stack) - 1
            if top >= 0 and stack[top] == '(':
                stack.pop()
            else:
                stack.append(s)
    
    if len(stack) == 0:
        print("YES")
    else:
        print("NO")

    stack.clear()

File: BOJ/nodejs/9012.js
/*
괄호
https://boj.kr/9012
 */
const start = new Date().getTime();

const input = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "input.txt")
  .toString()
  .split("\n");

const T = 1 * input[0];
const ans = new Array(T);
for (let t = 0; t < T; t++) {
  const PS = input[1 + t]; // 괄호 문자열(Parenthesis String, PS)

  const stack = new Array();
  let top = -1;
  for (let ch of PS) {
    if (ch === "(") {
      stack.push(ch);
    } else {
      if (stack[top] === "(") stack.pop();
      else stack.push(ch);
    }

    top = stack.length - 1;
  }

  ans[t] = top < 0 ? "YES" : "NO";
}

console.log(ans.join("\n"));

const end = new Date().getTime();
console.log("--------------------");
console.log(`Elapsed Time: ${(new Date().getTime() - start) / 1000}s`);
